### PR TITLE
Reader/Writer for winrt::hstring

### LINKED ISSUE #6166: Add ReadValue / WriteValue for winrt::hstring
>winrt::hstring [](start = 94, length = 14)

Since native module methods can support any parameter type which has ReadValue and WriteValue functions. we must fix this issue by adding the reader writer for winrt::hstring to avoid extra memory allocation due to the conversion.

_Originally posted by @vmoroz in https://github.com/microsoft/react-native-windows-samples/pull/280#discussion_r498542159_

### PR DESCRIPTION
**Pull request by BartoszKlonowski**
This pull request fixes #6166
The `winrt::hstring` is now handled separately when performing read/write operations on JS<->Native values passing.

---

This pull request implements the ReadValue and WriteValue functions as a separate overloads.
The benefit of this approach is that when passing the `winrt::hstring` value it is considered directly as this type, so there's no conversion required.
Without those overloads, the templated functions would catch the `is_convertible` and would pass it as the `std::string` or `std::wstring` which would cause yet another conversions after selecting and entering the overloaded constructor of `winrt::hstring`.

Changes introduced in this PR are:
* add `WriteValue` overloaded functions for both `winrt::hstring` and `winrt::param::hstring`
Even if using `winrt::param::` namespace should be avoided (see [hstring](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/strings#the-rationale-for-winrthstring-and-winrtparamhstring) docs) implementing explicit overloaded function for this type ensures that `winrt::param::` namespace won't be considered as `is_convertible` which would result in still inefficient conversions.
* add `ReadValue` function which takes the `winrt::hstring` type
There's no separation for `winrt::param::` anymore as there's no risk of falling into templated functions. Subset of `winrt::` namespace should be always considered as the best match for their parents overload.

For more details about the implementation please check the commit messages.

---

Unit tests are not modified in this PR as they are not affected by this implementation (no change in passing/failing ratio between before and after).

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/6704)

### CODE REVIEW

File: vnext/Microsoft.ReactNative.Cxx/JSValueReader.h [hunk 1]
```diff
@@ -183,6 +184,26 @@ inline void ReadValue(IJSValueReader const &reader, /*out*/ std::wstring &value)
...
+    case JSValueType::String:
+      value = reader.GetString();
+      break;
+    case JSValueType::Boolean:
+      value = reader.GetBoolean() ? L"true" : L"false";
+      break;
+    case JSValueType::Int64:
+      value = to_hstring(reader.GetInt64());
+      break;
+    case JSValueType::Double:
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative.Cxx/JSValueReader.h:198]: >case JSValueType::Double: [](start = 4, length = 25)

Please use the same code as the ReadValue for wstring above.

File: vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h [hunk 2]
```diff
@@ -23,6 +23,8 @@ template <class T, std::enable_if_t<std::is_convertible_v<T, std::string_view>,
 void WriteValue(IJSValueWriter const &writer, T const &value) noexcept;
 template <class T, std::enable_if_t<std::is_convertible_v<T, std::wstring_view>, int> = 1>
 void WriteValue(IJSValueWriter const &writer, T const &value) noexcept;
+void WriteValue(IJSValueWriter const &writer, winrt::hstring value) noexcept;
+void WriteValue(IJSValueWriter const &writer, const winrt::param::hstring value) noexcept;
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h:27]: I wonder why is it needed: the code for wstring_view above must just work.

Comment (vmoroz) [inline vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h:27]: >const winrt::param::hstring [](start = 46, length = 27)

Parameter must not be const on its own. It unnecessary restricts access to a variable passed by value.
Here and and in the line above we must use the 'const &' as we do in other places.
Though, first I want to understand if we need these two methods.

**Comment by vmoroz** [comment]
All new shared code must have supporting unit tests. Please add tests for the new methods.

**Review by vmoroz — CHANGES_REQUESTED** [review]
:clock1:

File: vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h (hunk 2 above)
Comment (asklar) [reply vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h:27]: this came from a comment you yourself had made in a PR :) https://github.com/microsoft/react-native-windows-samples/pull/280#discussion_r498542159

Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h:27]: Ah, I see the point: the wstring_view requires extra memory allocation. Never mind. :)

Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueWriter.h:27]: Please scratch the last question about if we need it or not: we need it.
The original question is still: please either remove the `const` or use `const &`

File: vnext/Microsoft.ReactNative.Cxx/JSValueReader.h (hunk 1 above)
Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueReader.h:198]: The dance with the string was implemented by FB to produce the right string there. Eventually we must fix it, but I want us to be consistent regardless which function we use.

Comment (asklar) [reply vnext/Microsoft.ReactNative.Cxx/JSValueReader.h:198]: is that code right though? it doesn't seem like it can reason over NaN or Inf values?

Comment (vmoroz) [reply vnext/Microsoft.ReactNative.Cxx/JSValueReader.h:198]: Agree. I do not like such value manipulations. We need to revisit it, but let's keep the results consistent for now.

**Comment by vmoroz** [comment]
> All new shared code must have supporting unit tests. Please add tests for the new methods.

Please add tests for the new methods in vnext\Microsoft.ReactNative.Cxx.UnitTests\JSValueReaderTest.cpp. The file name is a little bit confusing: it has tests for both reader and writer.

File: vnext/Microsoft.ReactNative.Cxx.UnitTests/JSValueReaderTest.cpp [hunk 3]
```diff
@@ -298,6 +298,7 @@ TEST_CLASS (JSValueReaderTest) {
       if (propertyName == L"StringValue1") {
         TestCheck(ReadValue<std::string>(reader) == "");
         TestCheck(ReadValue<std::wstring>(reader) == L"");
+        TestCheck(ReadValue<winrt::hstring>(reader) == L"");
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative.Cxx.UnitTests/JSValueReaderTest.cpp:301]: >ReadValue [](start = 18, length = 9)

Could you also add tests for the WriteValue? The goal is to have unit tests for all public methods.

**Comment by vmoroz** [comment]
> > All new shared code must have supporting unit tests. Please add tests for the new methods.
> 
> Please add tests for the new methods in vnext\Microsoft.ReactNative.Cxx.UnitTests\JSValueReaderTest.cpp. The file name is a little bit confusing: it has tests for both reader and writer.

It is nice to see the new tests for ReadValue. Could you also add tests for the WriteValue? We need to be sure that the new methods do not create any conflicts. For that, we should have the WriteValue tests that do not use an explicit type parameter, but rather just provide winrt::hstring and winrt::param:hstring values.

**Review by vmoroz — APPROVED** [review]
:shipit:

**Pull request merged**